File: packages/tolk/test/parity/simple_mul_i32/main.py
#!/usr/bin/env python3
"""Parity case: two int32 loads, multiply, store at index 0."""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from helpers import dump_stage7_program  # noqa: E402

from tinygrad.uop.ops import UOp, Ops, KernelInfo  # noqa: E402
from tinygrad.dtype import dtypes  # noqa: E402


def kernel():
    sink = UOp(Ops.SINK, dtypes.void, (), arg=KernelInfo())
    a = UOp.param(0, dtypes.int32, shape=(-1,))
    b = UOp.param(1, dtypes.int32, shape=(-1,))
    c = UOp.param(2, dtypes.int32, shape=(-1,))
    idx = UOp.const(dtypes.int, 0)
    idx_a = a.index(idx)
    ld_a = UOp(Ops.LOAD, dtypes.int32, (idx_a,))
    idx_b = b.index(idx)
    ld_b = UOp(Ops.LOAD, dtypes.int32, (idx_b,))
    mul = ld_a * ld_b
    idx_c = c.index(idx)
    store = UOp(Ops.STORE, dtypes.void, (idx_c, mul))
    return [sink, a, b, c, idx, idx_a, ld_a, idx_b, ld_b, mul, idx_c, store]


if __name__ == "__main__":
    dump_stage7_program(kernel(), os.path.dirname(os.path.abspath(__file__)))

File: packages/tolk/test/parity/simple_mul_i32/main.ml
(* Parity case: two int32 loads, multiply, store at index 0. *)

open Tolk_uop
module B = Program_spec_builder


let kernel () =
  let dt = Dtype.int32 in
  let ptr = dt in
  let b = B.create () in
  let p0 = B.emit b (Param { slot = 0; dtype = ptr }) in
  let p1 = B.emit b (Param { slot = 1; dtype = ptr }) in
  let p2 = B.emit b (Param { slot = 2; dtype = ptr }) in
  let c0 = B.emit b (Const { value = Const.int dt 0; dtype = dt }) in
  let idx0 =
    B.emit b (Index { ptr = p0; idxs = [ c0 ]; dtype = ptr })
  in
  let idx1 =
    B.emit b (Index { ptr = p1; idxs = [ c0 ]; dtype = ptr })
  in
  let ld0 = B.emit b (Load { src = idx0; alt = None; gate = None; dtype = dt }) in
  let ld1 = B.emit b (Load { src = idx1; alt = None; gate = None; dtype = dt }) in
  let prod =
    B.emit b (Binary { op = `Mul; lhs = ld0; rhs = ld1; dtype = dt })
  in
  let idx2 =
    B.emit b (Index { ptr = p2; idxs = [ c0 ]; dtype = ptr })
  in
  let _ = B.emit b (Store { dst = idx2; value = prod; gate = None }) in
  B.finish b

let () = Helpers.dump_stage7_program ~out_dir:Sys.argv.(1) (kernel ())
